### PR TITLE
Fix 5194: remove over-reaching additionalProperties:false post-processing

### PR DESCRIPTION
**Pull request by heath-freenome**
### Reasons for making this change

Fixes #5194, #5196, #3920

The additionalProperties:false post-processing block introduced in the #5176 fix built its key allowlist solely from the parent schema's `properties`, incorrectly deleting keys written by winning oneOf/anyOf branches and if/then branches. The block is removed entirely, restoring pre-PR behaviour where branch-granted keys survive omitExtraData filtering.

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [ ] **I'm adding or updating code**
  - [ ] I've added and/or updated tests. I've run `pnpm exec nx run-many --target=build --exclude=@rjsf/docs && pnpm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/utils/test/schema/omitExtraDataTest.ts [hunk 1]
```diff
@@ -1189,6 +1190,55 @@ export default function omitExtraDataTest(testValidator: TestValidatorType) {
...
+          type: 'object',
+          additionalProperties: false,
+          if: { type: 'object' } as any,
+          then: { properties: { extra: { type: 'string' } } },
+        };
+        const formData = { extra: 'keep' };
+        testValidator.setReturnValues({ isValid: [true] });
+        const result = omitExtraData(testValidator, schema, schema, formData);
+        expect(result).toEqual({ extra: 'keep' });
+      });
```
Comment (x0k) [inline packages/utils/test/schema/omitExtraDataTest.ts]: Why was this behavior chosen instead of "Prune additional properties"?

The values obtained do not pass validation: [case1](https://x0k.dev/svelte-jsonschema-form/playground3/?page=v#N4IgzgxgFgpgtgQxALhMAOgOwATYC4CeADjMtgOQD2ARgFYwR7kA0WuRATpSR3gJYwwZYPmKlsIwiTLkweDn0wBzctgC+61jmxEEePDA6YAClx79BwigD0iAfXJWp42fMUr1mttgQATX3z8lJgIADam3IYWQtgAZmFgMFq4fLFWnJG8AjGSYlYQwXIyAIKqGuXJ+LCY6WZR2VYwAB7yCE55FHIKymWealpqIMwgikQArngoaKLSFKXMOnaxAqG+MgDWMDBELNjNrRtbO+pDIJQT45OoANreGNq46CC6eFBPZNcAupWPIHCCYAQShg72wTzgYzk2AAcgB5AAq2CgCAAbjAfP5AnxgmEdHUsoInt41FhPqcUWE+L49JQOChQCF-lMELQUQAOU6+DgIWJ4ABMAAZBSh4qFEsNOAxKHAiHxQjBfCKEjA1GogA), [case2](https://x0k.dev/svelte-jsonschema-form/playground3/?page=v#N4IgzgxgFgpgtgQxALhMAOgOwATYC4CeADjMtgOQD2ARgFYwR7kA0WuCAJhwJZ7eWYEAGwAKAJ0okxfGGDIAzYWBisc2bvLLB8xUhRr1G5bAF9VuPLExbsRCVJlzs2mAA88YhDcIky5MB7cmADmxiampqomIMwgQUQArngoaNhuHl4UANYwMERhMSCUSYnJqADabM5VuOggRAiWdWTlALrmuNh1cLJgCMEwzV0gcAkB2AByAPIAKthQCABuMNicPHwCwrb2MNLcsnVVJlithYvC3ByNlGIooII9KQi0iwAchRye8ngATAAM-xQiiEylidgYlDgRG4QhgHCBShgJhMQA)

Comment (heath-freenome) [reply packages/utils/test/schema/omitExtraDataTest.ts]: Good question — there are two different versions of "restoring the adaptation" worth distinguishing:

**The original (buggy) adaptation** — what this PR removes — built the allowed-key set from the parent schema's `properties` only, so it wiped out keys written by winning `oneOf`/`anyOf` and `if`/`then` branches. That's the root cause of #5194, so restoring it exactly makes things worse.

**A corrected SJSF-style adaptation** that tracks *all* keys written by any branch during processing (parent `properties` + `patternProperties` + branch-granted keys) and only prunes what isn't covered by any of them: for your two playground examples the runtime result would be **identical to this PR** — `{ type: 'A', p_field: 'keep', extra: 'keep' }` — because `extra` is branch-granted. The data would still fail strict AJV validation for the same reason: in JSON Schema draft-07, `additionalProperties: false` only sees `properties` and `patternProperties` at the same schema level; it does not retroactively count `then`-branch properties as "known."

The choice to preserve branch-granted keys is intentional: if `type: 'A'` reveals an `extra` field and the user fills it in, stripping it on submit would be a surprising regression even if a strict validator would flag it. The spec-correct solution for cross-keyword property awareness is `unevaluatedProperties` (draft 2019-09+), not `additionalProperties: false`.

That said, if you think `omitExtraData` should enforce stricter pruning here and you have a concrete approach that handles all the branch-composition edge cases, I'm open to it — it's a real trade-off.

Comment (x0k) [reply packages/utils/test/schema/omitExtraDataTest.ts]: >  for your two playground examples the runtime result would be identical to this PR 

[omit1](https://x0k.dev/svelte-jsonschema-form/playground3/?page=o#N4IgzgxgFgpgtgQxALhMAOgOwATYC4CeADjMtgOQD2ARgFYwR7kA0WuRATpSR3gJYwwZYPmKlsIwiTLkweDn0wBzctgC+61jmxEEePDA6YAClx79BwigD0iAfXJWp42fMUr1mttgQATX3z8lJgIADam3IYWQtgAZmFgMFq4fLFWnJG8AjGSYlYQwXIyAIKqGuXJ+LCY6WZR2VYwAB7yCE55FHIKymWealpqIMwgikQArngoaKLSFKXMOnaxAqG+MgDWMDBELNjNrRtbO+pDIJQT45OoGNroIM53ZHfFd5V39sswq4-Yd5vbdywgzUQA), [omit2](https://x0k.dev/svelte-jsonschema-form/playground3/?page=o#N4IgzgxgFgpgtgQxALhMAOgOwATYC4CeADjMtgOQD2ARgFYwR7kA0WuCAJhwJZ7eWYEAGwAKAJ0okxfGGDIAzYWBisc2bvLLB8xUhRr1G5bAF9VuPLExbsRCVJlzs2mAA88YhDcIky5MB7cmADmxiampqomIMwgQUQArngoaNhuHl4UANYwMERhMSCUSYnJqMDRJkA)

This is handled by [this code](https://github.com/x0k/svelte-jsonschema-form/blob/3ee66f60d4f9ea0d08b93857966bdcc53b973648/packages/form/src/omit-extra-data.ts#L311-L322)

Comment (heath-freenome) [reply packages/utils/test/schema/omitExtraDataTest.ts]: I integrated that fix. And then fixed a bug.

Comment (x0k) [reply packages/utils/test/schema/omitExtraDataTest.ts]: The validator doesn't like this  [demo](https://x0k.dev/svelte-jsonschema-form/playground3/?page=v#N4IgzgxgFgpgtgQxALhMAOgOwATYC4CeADjMtgOQD2ARgFYwR7kA0WuRATpSR3gJYwwZYNkwI4pbCMIky5MHg59MAc3LYAvptY5sCACb6+-SmIA2ABS49+gsgDMEZsDB259MEpg+YIAoVJsuKLikhi6wdic3DC8-sJ6KmH4xJLkmACucNSx6loabsEFQcWYGiDMIMpEGXgoaCEScgCCZnwQMCyJkgDMAAyaFSCUtTV1qADaQeGR6CBECHhQc2QTALqFuHMSYGAISSvY2xkK2AByAPIAKthQCABuMHqGxnymTlHWsbZgcyVYayG9ycfH0i0oHBQoDEEnqCFo9wAHEN9BwEPY8AAmPrYlCOZyueYcBiUOBEPhmGD6PFOFwaDRAA).
I think this might be one of the reasons to switch to the new version of the JSON schema with `unevaluatedProperties`

Comment (heath-freenome) [reply packages/utils/test/schema/omitExtraDataTest.ts]: @x0k which validator? There are 3 now

Comment (x0k) [reply packages/utils/test/schema/omitExtraDataTest.ts]: All of them 
image

File: packages/utils/test/schema/omitExtraDataTest.ts [hunk 2]
```diff
@@ -1189,6 +1283,56 @@ export default function omitExtraDataTest(testValidator: TestValidatorType) {
...
+        // is pruned. 'p_field' matches '^p_' and is retained.
+        const schema: RJSFSchema = {
+          type: 'object',
+          properties: { type: { type: 'string' } },
+          patternProperties: { '^p_': { type: 'string' } },
+          additionalProperties: false,
+          if: { properties: { type: { const: 'A' } } },
+          then: { properties: { extra: { type: 'string' } } },
+        };
+        const formData = { type: 'A', p_field: 'keep', extra: 'keep' };
```
Comment (nickgros) [inline packages/utils/test/schema/omitExtraDataTest.ts]: Change this value to 'drop' to match the intent of the test

```suggestion
        const formData = { type: 'A', p_field: 'keep', extra: 'drop' };
```

File: packages/utils/test/schema/omitExtraDataTest.ts [hunk 3]
```diff
@@ -1236,12 +1391,13 @@ export default function omitExtraDataTest(testValidator: TestValidatorType) {
...
-        // anyOf sees [] as empty → applies all branches → returns [] as target.
-        // The outer type:'array' branch then runs handleArray with that existing [] as target,
-        // hitting the Array.isArray(target) ? target : [] true-branch.
+        // anyOf sees [] as empty → applies all branches.
+        // The branch has type:'array' so omit() returns [] as the filtered value.
+        // The outer type:'array' block then runs handleArray with that existing [] as target,
+        // hitting the Array.isArray(filtered) ? filtered : [] true-branch (filtered is an array).
         const schema: RJSFSchema = {
           type: 'array',
-          anyOf: [{ items: { type: 'string' } }, { items: { type: 'number' } }],
```
Comment (nickgros) [inline packages/utils/test/schema/omitExtraDataTest.ts:1254]: Thanks for fixing this test to match the description. Do we still need to have a test that matches the original test logic (or do we have one I am not seeing and this was a duplicate)?

File: packages/utils/src/schema/omitExtraData.ts [hunk 4]
```diff
@@ -424,26 +429,30 @@ export default function omitExtraData<
...
    * @param schemaDef - The schema (or boolean shorthand) to filter `source` against
    * @param source - The raw form data value to filter
    * @param [target] - An optional accumulator carrying results from prior oneOf/anyOf processing
+   * @param [materializeSource=true] - When false a permissive schema (`true`/`{}`) returns `target`
+   *   rather than `source`, preventing compositional branches from aliasing the accumulator to `source`
    * @returns - The filtered value, or `undefined` when the schema rejects the value
    */
-  function omit(schemaDef: S | boolean, source: unknown, target?: unknown): unknown {
+  function omit(schemaDef: S | boolean, source: unknown, target?: unknown, materializeSource = true): unknown {
     if (source === undefined || schemaDef === false) {
```
Comment (nickgros) [inline packages/utils/src/schema/omitExtraData.ts]: What do you think about `useSourceAsFallback` instead of `materializeSource`?

**Review by nickgros — APPROVED** [review]
Need a clean build, but generally LGTM with some minor requests. Please let me know if I should re-review once @x0k 's feedback is addressed

**Comment by heath-freenome** [comment]
@x0k Any thoughts on the current state of this PR? More precisely do you have an object to it getting merged?

**Comment by x0k** [comment]
I still don't understand why the `preserves keys added by a dependency sub-schema even when parent has additionalProperties:false` case was added if its result doesn't pass validation [ajv](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJuYW1lIjoiQWxpY2UiLCJhZ2UiOjMwfSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7Im5hbWUiOnsidHlwZSI6InN0cmluZyJ9fSwiYWRkaXRpb25hbFByb3BlcnRpZXMiOmZhbHNlLCJkZXBlbmRlbmNpZXMiOnsibmFtZSI6eyJwcm9wZXJ0aWVzIjp7ImFnZSI6eyJ0eXBlIjoibnVtYmVyIn19fX19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJzaG93RXJyb3JMaXN0IjoidG9wIiwidmFsaWRhdGUiOmZhbHNlLCJkaXNhYmxlZCI6ZmFsc2UsIm5vSHRtbDVWYWxpZGF0ZSI6ZmFsc2UsInJlYWRvbmx5IjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZSwiZXhwZXJpbWVudGFsX2NvbXBvbmVudFVwZGF0ZVN0cmF0ZWd5IjoiY3VzdG9tRGVlcCIsImV4cGVyaW1lbnRhbF9kZWZhdWx0Rm9ybVN0YXRlQmVoYXZpb3IiOnsiYXJyYXlNaW5JdGVtcyI6InBvcHVsYXRlIiwiYWxsT2YiOiJza2lwRGVmYXVsdHMiLCJjb25zdEFzRGVmYXVsdHMiOiJhbHdheXMiLCJlbXB0eU9iamVjdEZpZWxkcyI6InBvcHVsYXRlQWxsRGVmYXVsdHMiLCJtZXJnZURlZmF1bHRzSW50b0Zvcm1EYXRhIjoidXNlRm9ybURhdGFJZlByZXNlbnQiLCJuZXN0ZWREZWZhdWx0c1ByZWNlZGVuY2UiOiJkZXNjZW5kYW50V2lucyJ9LCJ1c2VGYWxsYmFja0ZpZWxkIjpmYWxzZSwidXNlRmFsbGJhY2tVaUZvclVuc3VwcG9ydGVkVHlwZSI6ZmFsc2UsImxpdmVWYWxpZGF0ZSI6ZmFsc2UsImlkUHJlZml4Ijoicm9vdCIsImlkU2VwYXJhdG9yIjoiXyJ9LCJ2YWxpZGF0b3IiOiJBSlY4Iiwic2FtcGxlTmFtZSI6IkJsYW5rIn0=), [ata](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJuYW1lIjoiQWxpY2UiLCJhZ2UiOjMwfSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7Im5hbWUiOnsidHlwZSI6InN0cmluZyJ9fSwiYWRkaXRpb25hbFByb3BlcnRpZXMiOmZhbHNlLCJkZXBlbmRlbmNpZXMiOnsibmFtZSI6eyJwcm9wZXJ0aWVzIjp7ImFnZSI6eyJ0eXBlIjoibnVtYmVyIn19fX19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJzaG93RXJyb3JMaXN0IjoidG9wIiwidmFsaWRhdGUiOmZhbHNlLCJkaXNhYmxlZCI6ZmFsc2UsIm5vSHRtbDVWYWxpZGF0ZSI6ZmFsc2UsInJlYWRvbmx5IjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZSwiZXhwZXJpbWVudGFsX2NvbXBvbmVudFVwZGF0ZVN0cmF0ZWd5IjoiY3VzdG9tRGVlcCIsImV4cGVyaW1lbnRhbF9kZWZhdWx0Rm9ybVN0YXRlQmVoYXZpb3IiOnsiYXJyYXlNaW5JdGVtcyI6InBvcHVsYXRlIiwiYWxsT2YiOiJza2lwRGVmYXVsdHMiLCJjb25zdEFzRGVmYXVsdHMiOiJhbHdheXMiLCJlbXB0eU9iamVjdEZpZWxkcyI6InBvcHVsYXRlQWxsRGVmYXVsdHMiLCJtZXJnZURlZmF1bHRzSW50b0Zvcm1EYXRhIjoidXNlRm9ybURhdGFJZlByZXNlbnQiLCJuZXN0ZWREZWZhdWx0c1ByZWNlZGVuY2UiOiJkZXNjZW5kYW50V2lucyJ9LCJ1c2VGYWxsYmFja0ZpZWxkIjpmYWxzZSwidXNlRmFsbGJhY2tVaUZvclVuc3VwcG9ydGVkVHlwZSI6ZmFsc2UsImxpdmVWYWxpZGF0ZSI6ZmFsc2UsImlkUHJlZml4Ijoicm9vdCIsImlkU2VwYXJhdG9yIjoiXyJ9LCJ2YWxpZGF0b3IiOiJBVEEiLCJzYW1wbGVOYW1lIjoiQmxhbmsifQ==), [cfworker](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJuYW1lIjoiQWxpY2UiLCJhZ2UiOjMwfSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7Im5hbWUiOnsidHlwZSI6InN0cmluZyJ9fSwiYWRkaXRpb25hbFByb3BlcnRpZXMiOmZhbHNlLCJkZXBlbmRlbmNpZXMiOnsibmFtZSI6eyJwcm9wZXJ0aWVzIjp7ImFnZSI6eyJ0eXBlIjoibnVtYmVyIn19fX19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJzaG93RXJyb3JMaXN0IjoidG9wIiwidmFsaWRhdGUiOmZhbHNlLCJkaXNhYmxlZCI6ZmFsc2UsIm5vSHRtbDVWYWxpZGF0ZSI6ZmFsc2UsInJlYWRvbmx5IjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZSwiZXhwZXJpbWVudGFsX2NvbXBvbmVudFVwZGF0ZVN0cmF0ZWd5IjoiY3VzdG9tRGVlcCIsImV4cGVyaW1lbnRhbF9kZWZhdWx0Rm9ybVN0YXRlQmVoYXZpb3IiOnsiYXJyYXlNaW5JdGVtcyI6InBvcHVsYXRlIiwiYWxsT2YiOiJza2lwRGVmYXVsdHMiLCJjb25zdEFzRGVmYXVsdHMiOiJhbHdheXMiLCJlbXB0eU9iamVjdEZpZWxkcyI6InBvcHVsYXRlQWxsRGVmYXVsdHMiLCJtZXJnZURlZmF1bHRzSW50b0Zvcm1EYXRhIjoidXNlRm9ybURhdGFJZlByZXNlbnQiLCJuZXN0ZWREZWZhdWx0c1ByZWNlZGVuY2UiOiJkZXNjZW5kYW50V2lucyJ9LCJ1c2VGYWxsYmFja0ZpZWxkIjpmYWxzZSwidXNlRmFsbGJhY2tVaUZvclVuc3VwcG9ydGVkVHlwZSI6ZmFsc2UsImxpdmVWYWxpZGF0ZSI6ZmFsc2UsImlkUHJlZml4Ijoicm9vdCIsImlkU2VwYXJhdG9yIjoiXyJ9LCJ2YWxpZGF0b3IiOiJDRldvcmtlciAoRHJhZnQgNykiLCJzYW1wbGVOYW1lIjoiQmxhbmsifQ==) (click raw validation).

**Comment by heath-freenome** [comment]
@x0k You're technically correct about the JSON Schema spec: `additionalProperties` is evaluated against `properties` and `patternProperties` at the **same schema level only**. `dependencies` subschemas do not merge their `properties` into the parent schema's allowed-property set — they're evaluated as a separate overlay. So `{ name: "Alice", age: 30 }` is strictly invalid against that schema regardless of what `name` holds.

But `omitExtraData` is not a validator. Its contract is: *strip data for fields RJSF did not render*. When `name` is filled in, RJSF activates the `dependencies` sub-schema and **renders an `age` input**. The user types into it. If `omitExtraData` then deletes `age`, the user has lost data they explicitly entered into a visible field — a fundamental UX contract violation.

The PR's pipeline makes the ordering explicit:

```
source → handleObject → handleConditions → [post-prune for additionalProperties:false] → handleDependencies → result
```

The post-prune step fires **after** conditional branches (`if/then`, `oneOf`, `anyOf`) but **before** `handleDependencies`. Keys that only survive because of a conditional branch *are* pruned when `additionalProperties: false` is set. Keys added by `dependencies` sub-schemas are not, because those sub-schemas directly cause RJSF to render new fields.

The schema in the test is an anti-pattern: combining `additionalProperties: false` with `dependencies` that introduce new properties creates a schema where the only valid data is data the rendered form can't produce. RJSF has to pick a side — strip rendered-field data (bad UX) or preserve it (bad strict-spec compliance). For a form library, preserving rendered-field data is the right call.

If strict schema compliance is the goal, the fix is in the schema: move `age` into the top-level `properties` (with the dependency sub-schema adding only constraints), or replace `dependencies` with `if/then`. With `if/then`, the PR's post-prune step *does* delete `age` when `additionalProperties: false` is set, so the output would actually validate.

**Comment by x0k** [comment]
> Its contract is: strip data for fields RJSF did not render

This may be one source of the misunderstanding, since I had a different intent in mind for this function:

Remove data that is not declared by the JSON Schema (with no renderer-specific or other logic involved - just a clear standard to follow).

> a fundamental UX contract violation.

However, this can result in a form that cannot be submitted unless these values are removed manually, or may be impossible to submit at all if those fields are declared as required.

That doesn't seem like good UX either. Instead of choosing between two bad options, we could:

* Add a dev-time check for schemas that are incompatible with RJSF and provide suggestions for how to change the schema.
* Explicitly support the workarounds used to achieve the equivalent result.

```js
{
  type: 'object',
  properties: {
    name: { type: 'string' },
    age: {}
  },
  additionalProperties: false,
  dependencies: {
    name: {
      properties: { age: { type: 'number' } }
    }
  }
}
```

```js
{
  type: 'object',
  oneOf: [
    {
      properties: { name: { type: 'string' } },
      additionalProperties: false
    },
    {
      properties: {
        name: { type: 'string' },
        age: { type: 'number' }
      },
      required: ['name'],
      additionalProperties: false
    }
  ]
}
```
